### PR TITLE
Image Segmentation on Android demo app

### PR DESCRIPTION
**Pull request by jeffxtang**

### CODE REVIEW

File: ImageSegmentation/app/build.gradle [hunk 1]
```diff
@@ -0,0 +1,35 @@
+apply plugin: 'com.android.application'
+
+android {
+    compileSdkVersion 30
+    buildToolsVersion "30.0.2"
+
+    defaultConfig {
+        applicationId "com.example.imagesegmentation"
```
Comment (IvanKobzarev) [inline ImageSegmentation/app/build.gradle:8]: For the first demo we used package org.pytorch.demo
Could it be also here, something like `org.pytorch.demo.imagesegmentation` ?

File: ImageSegmentation/app/proguard-rules.pro [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Add project specific ProGuard rules here.
```
Comment (IvanKobzarev) [inline ImageSegmentation/app/proguard-rules.pro:1]: This file is all commented out, maybe remove this file and remove it from build.gradle?

File: ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+import java.io.OutputStream;
+import java.util.Map;
+
+public class MainActivity extends AppCompatActivity implements Runnable {
+    private ImageView imageView;
+    private Button buttonSegment;
+    private ProgressBar progressBar;
+    private Bitmap bitmap = null;
+    private Module module = null;
+    private String imagename = "deeplab.jpg";
```
Comment (IvanKobzarev) [inline ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java:33]: style nit: on your choice, for previous demo apps member fields were prefixed with m,
`mImageView`, `mButtonSegment` etc.

File: ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                progressBar.setVisibility(ProgressBar.VISIBLE);
+                buttonSegment.setText(getString(R.string.run_model));
+
+                Thread thread = new Thread(MainActivity.this);
+                thread.start();
+            }
+        });
+
+        try {
+            module = Module.load(MainActivity.assetFilePath(getApplicationContext(), "deeplabv3_scripted.pt"));
```
Comment (IvanKobzarev) [inline ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java:107]: `MainActivity.assetFilePath` function can be removed if to switch to 
```
PyTorchAndroid.loadModuleFromAsset(getAssets(), "deeplabv3_scripted.pt")
```
as in https://github.com/pytorch/pytorch/blob/master/android/test_app/app/src/main/java/org/pytorch/testapp/MainActivity.java#L132

File: ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            Log.e("ImageSegmentation", "Error reading assets", e);
+            finish();
+        }
+
+    }
+
+    @Override
+    public void run() {
+        final Tensor inputTensor = TensorImageUtils.bitmapToFloat32Tensor(bitmap,
+                TensorImageUtils.TORCHVISION_NORM_MEAN_RGB, TensorImageUtils.TORCHVISION_NORM_STD_RGB);
```
Comment (IvanKobzarev) [inline ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java:118]: does that model uses completely the same normalization constants as for torchvision pretrained models?

File: ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        final float[] inputs = inputTensor.getDataAsFloatArray();
+        Map<String, IValue> outTensors = module.forward(IValue.from(inputTensor)).toDictStringKey();
+        final Tensor outputTensor = outTensors.get("out").toTensor();
+        final float[] scores = outputTensor.getDataAsFloatArray();
+        int width = bitmap.getWidth();
+        int height = bitmap.getHeight();
+        int[] intValues = new int[width * height];
+//        for (int i = 0; i < intValues.length; i++) {
+//            intValues[i] = 0xFFFFFFFF;
+//        }
```
Comment (IvanKobzarev) [inline ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java:128]: clean?

File: ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        for (int j = 0; j < width; j++) {
+            for (int k = 0; k < height; k++) {
+                int maxi = 0, maxj = 0, maxk = 0;
+                double maxnum = -100000.0;
```
Comment (IvanKobzarev) [inline ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java:132]: nit, magic number `-100000.0` can be replaced to `-Double.MAX_VALUE`

File: ImageSegmentation/deeplabv3.py [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import torch
+
+model = torch.hub.load('pytorch/vision:v0.7.0', 'deeplabv3_resnet50', pretrained=True)
+model.eval()
+
+scriptedm = torch.jit.script(model)
+torch.jit.save(scriptedm, "deeplabv3_scripted.pt")
```
Comment (IvanKobzarev) [inline ImageSegmentation/deeplabv3.py:7]: maybe to store it directly to assets folder?
```
torch.jit.save(scriptedm, "app/src/main/assets/deeplabv3_scripted.pt")
```

**Review by IvanKobzarev — APPROVED** [review]

File: ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java (hunk 5 above)
Comment (jeffxtang) [reply ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java:118]: yes

File: ImageSegmentation/deeplabv3.py (hunk 8 above)
Comment (jeffxtang) [reply ImageSegmentation/deeplabv3.py:7]: yes yours is better. but i'll do it in the future when updating the accompanying tutorial and the readme as this is mentioned in a few places.

File: ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java (hunk 4 above)
Comment (jeffxtang) [reply ImageSegmentation/app/src/main/java/org/pytorch/imagesegmentation/MainActivity.java:107]: cool.. but i'll use this method in the next demo app (and update it for this app too at that time).
